Userland/Libraries/LibCore/Group.cpp

```diff
@@ -0,0 +1,125 @@
+}
+
+ErrorOr<Optional<Group>> Group::from_gid(gid_t group_id)
+{
+    auto group = TRY(Core::System::getgrgid(group_id));
+    if (!group.has_value()) {
+        return Optional<Group> {};
+    }
+
+    return TRY(from_libc_group(group.value()));
+}
+
+#ifndef AK_OS_MACOS
+ErrorOr<void> Group::add_group(Group& group)
+{
+    if (group.group_name().is_empty())
+        return Error::from_string_literal("Group name can not be empty.");
+
+    // A quick sanity check on group name
+    if (strpbrk(group.group_name().characters(), "\\/!@#$%^&*()~+=`:\n"))
+        return Error::from_string_literal("Group name has invalid characters.");
+
+    // Disallow names starting with '_', '-' or other non-alpha characters.
+    if (group.group_name().starts_with('_') || group.group_name().starts_with('-') || !is_ascii_alpha(group.group_name().characters()[0]))
+        return Error::from_string_literal("Group name has invalid characters.");
+
+    // Verify group name does not already exist
+    if (TRY(Core::System::getgrnam(group.group_name())).has_value())
+        return Error::from_string_literal("Group name already exists.");
+
+    // Sort out the group id for the group
+    if (group.group_id() > 0) {
+        if (TRY(Core::System::getgrgid(group.group_id())).has_value())
+            return Error::from_string_literal("Group ID already exists.");
+    } else {
+        unsigned int group_id = 100;
+        while (true) {
+            auto gr = TRY(Core::System::getgrgid(group_id));
+            if (!gr.has_value())
+                break;
+            group_id++;
+        }
+        group.set_group_id(group_id);
+    }
+
+    auto gr = TRY(group.to_libc_group());
+
+    errno = 0;
+    FILE* file = fopen("/etc/group", "a");
+    if (errno)
+        return Error::from_errno(errno);
+
+    int rc = putgrent(&gr, file);
+    if (rc < 0)
+        return Error::from_errno(errno);
+
+    fclose(file);
+
+    return {};
+}
+#endif
+
+Group::Group(String group_name, unsigned int group_id, Vector<String> group_members)
+    : m_group_name(move(group_name))
+    , m_group_id(group_id)
+    , m_group_members(move(group_members))
+{
+}
+
+// NOTE: struct group returned from this function cannot outlive an instance of Group.
+ErrorOr<struct group> Group::to_libc_group()
+{
+    struct group gr;
+    gr.gr_name = const_cast<char*>(m_group_name.characters());
+    gr.gr_passwd = const_cast<char*>("x");
+    gr.gr_gid = m_group_id;
+    gr.gr_mem = nullptr;
+
+    // FIXME: A better solution would surely be not using a static here
+    // NOTE: This now means that there cannot be multiple struct groups at the same time, because only one gr.gr_mem can ever be valid at the same time.
+    // NOTE: Not using a static here would result in gr.gr_mem being freed up on exit from this function.
+    static Vector<char*> members;
+    members.clear_with_capacity();
+    if (m_group_members.size() > 0) {
+        TRY(members.try_ensure_capacity(group_members().size() + 1));
```

Comment:
You missed a call to `group_members()`, just make this `m_group_members` to be consistent :^)

Reply:
Done:)